File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/DerivedTerminalGrammarAutomata.java
package titan.ast.runtime;

import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import titan.ast.runtime.AstGeneratorResult.TokensResult;
import titan.ast.runtime.DerivedTerminalGrammarAutomataData.RootTerminalGrammarMap;

/**
 * .
 *
 * @author tian wei jun
 */
class DerivedTerminalGrammarAutomata extends DfaTokenAutomata {
  Map<Grammar, HashMap<String, Grammar>> rootTerminalGrammarMap;

  /**
   * 初始化oneTokenStringBuilder、eof、dfa.
   *
   * @param derivedTerminalGrammarAutomataData 关键字有限自动机
   * @param dfa 识别token的 确定有限状态自动机
   */
  DerivedTerminalGrammarAutomata(
      DerivedTerminalGrammarAutomataData derivedTerminalGrammarAutomataData, TokenDfa dfa) {
    super(dfa);
    rootTerminalGrammarMap = new HashMap<>(derivedTerminalGrammarAutomataData.count);
    for (RootTerminalGrammarMap map : derivedTerminalGrammarAutomataData.rootTerminalGrammarMaps) {
      rootTerminalGrammarMap.put(map.rootTerminalGrammar, map.textTerminalMap);
    }
  }

  @Override
  public TokensResult buildToken(InputStream byteInputStream) {
    TokensResult tokensResult = super.buildToken(byteInputStream);
    if (tokensResult.isOk()) {
      buildToken(tokensResult.getOkData());
    }
    return tokensResult;
  }

  List<Token> buildToken(List<Token> tokens) {
    for (Token token : tokens) {
      HashMap<String, Grammar> textTerminalMap = rootTerminalGrammarMap.get(token.terminal);
      if (textTerminalMap != null) {
        Grammar terminal = textTerminalMap.get(token.text);
        if (null != terminal) {
          token.terminal = terminal;
        }
      }
    }
    return tokens;
  }
}

File: runtime/cpp/titan-ast-runtime-lib/DerivedTerminalGrammarAutomata.cpp
//
// Created by tian wei jun on 2022/11/25 0025.
//

#include "DerivedTerminalGrammarAutomata.h"

DerivedTerminalGrammarAutomata::DerivedTerminalGrammarAutomata(
    const DerivedTerminalGrammarAutomataData
        *derivedTerminalGrammarAutomataData,
    const TokenDfa *tokenDfa)
    : DfaTokenAutomata(tokenDfa) {
  rootTerminalGrammarMap = std::unordered_map<
      const Grammar *,
      const std::unordered_map<std::string *, Grammar *, TextTerminalMapHash,
                         TextTerminalMapEq> *,
      PtrGrammarContentHash, PtrGrammarContentEq>(
      derivedTerminalGrammarAutomataData->count);
  for (auto &map :
       derivedTerminalGrammarAutomataData->rootTerminalGrammarMaps) {
    const std::pair<const Grammar *,
              const std::unordered_map<std::string *, Grammar *, TextTerminalMapHash,
                                 TextTerminalMapEq> *>
        pair(map.rootTerminalGrammar, &map.textTerminalMap);
    rootTerminalGrammarMap.insert(pair);
  }
}

// dfa delete by PersistentObject.tokenDfa
// derivedTerminalGrammarAutomataData delete by
// PersistentObject.derivedTerminalGrammarAutomataData tokens delete by caller
DerivedTerminalGrammarAutomata::~DerivedTerminalGrammarAutomata() = default;

TokensResult *
DerivedTerminalGrammarAutomata::buildToken(const std::string *sourceFilePath) {
  auto tokensResult = DfaTokenAutomata::buildToken(sourceFilePath);
  if (tokensResult->isOk()) {
    buildTokenByDerivedTerminalGrammarAutomata(tokensResult->getOkData());
  }
  return tokensResult;
}
std::vector<Token *> *DerivedTerminalGrammarAutomata::buildTokenByDerivedTerminalGrammarAutomata(
    std::vector<Token *> *tokens) const {
  for (auto token : *tokens) {
    auto rootTerminalGrammarMapIt = rootTerminalGrammarMap.find(&token->terminal);
    if (rootTerminalGrammarMapIt != rootTerminalGrammarMap.end()) {
      auto textTerminalMap = rootTerminalGrammarMapIt->second;
      auto terminalIt = textTerminalMap->find(&token->text);
      if (terminalIt != textTerminalMap->end()) {
        auto terminal = terminalIt->second;
        token->terminal = *terminal;
      }
    }
  }
  return tokens;
}




